File: src/health_bridge/contract/batch_v1.py
import re
from datetime import UTC, datetime
from typing import Annotated, ClassVar, Final, Literal, Self, TypeAlias

from pydantic import (
    AfterValidator,
    BaseModel,
    BeforeValidator,
    ConfigDict,
    Field,
    StringConstraints,
    field_validator,
    model_validator,
)
from pydantic_core import PydanticCustomError

SCHEMA_MAJOR_ERROR_TYPE: Final = "unsupported_schema_major"
SCHEMA_MAJOR_ERROR_MESSAGE: Final = "schema_version must match major version 1"
EXPLICIT_NULL_ERROR_TYPE: Final = "explicit_null_not_allowed"
EXPLICIT_NULL_ERROR_MESSAGE: Final = "field may be omitted but cannot be null"
ALIASES_UNIQUE_ERROR_TYPE: Final = "aliases_not_unique"
ALIASES_UNIQUE_ERROR_MESSAGE: Final = "aliases must be unique"
SCHEMA_VERSION_PATTERN: Final = re.compile(r"^1\.\d+\.\d+$")
UTC_TIMESTAMP_PATTERN: Final = r"^\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}Z$"
SYNTHETIC_SOURCE_PATTERN: Final = r"^(synthetic|apple_health)\.[a-z0-9_.-]+$"
SYNTHETIC_RECORD_PATTERN: Final = r"^(synthetic|hk)-[a-z0-9-]+$"
TYPE_CODE_PATTERN: Final = r"^[a-z][a-z0-9_]*$"


def validate_utc_timestamp(value: str) -> str:
    try:
        _ = datetime.strptime(value, "%Y-%m-%dT%H:%M:%SZ").replace(tzinfo=UTC)
    except ValueError as exc:
        error_type = "invalid_utc_timestamp"
        error_message = "timestamp must be a valid UTC timestamp"
        raise PydanticCustomError(
            error_type,
            error_message,
        ) from exc
    return value


def utc_datetime(value: str) -> datetime:
    return datetime.strptime(value, "%Y-%m-%dT%H:%M:%SZ").replace(tzinfo=UTC)


def validate_order(start_time: str, end_time: str) -> None:
    if utc_datetime(start_time) > utc_datetime(end_time):
        message = "start_time must not be after end_time"
        raise ValueError(message)


UtcTimestamp: TypeAlias = Annotated[
    str,
    StringConstraints(pattern=UTC_TIMESTAMP_PATTERN),
    AfterValidator(validate_utc_timestamp),
]
SyntheticSourceKey: TypeAlias = Annotated[
    str,
    StringConstraints(pattern=SYNTHETIC_SOURCE_PATTERN),
]
SyntheticRecordId: TypeAlias = Annotated[
    str,
    StringConstraints(pattern=SYNTHETIC_RECORD_PATTERN),
]
TypeCode: TypeAlias = Annotated[str, StringConstraints(pattern=TYPE_CODE_PATTERN)]
NonEmptyString: TypeAlias = Annotated[str, StringConstraints(min_length=1)]
NonNegativeFloat: TypeAlias = Annotated[float, Field(ge=0)]
NonNegativeInt: TypeAlias = Annotated[int, Field(ge=0)]
JsonNumber: TypeAlias = int | float


def reject_explicit_null_string(value: str | None) -> str | None:
    if value is None:
        raise PydanticCustomError(EXPLICIT_NULL_ERROR_TYPE, EXPLICIT_NULL_ERROR_MESSAGE)
    return value


def reject_explicit_null_number(value: JsonNumber | None) -> JsonNumber | None:
    if value is None:
        raise PydanticCustomError(EXPLICIT_NULL_ERROR_TYPE, EXPLICIT_NULL_ERROR_MESSAGE)
    return value


OmittableString: TypeAlias = Annotated[
    str | None,
    BeforeValidator(reject_explicit_null_string),
]
OmittableNonNegativeFloat: TypeAlias = Annotated[
    NonNegativeFloat | None,
    BeforeValidator(reject_explicit_null_number),
]


class StrictModel(BaseModel):
    model_config: ClassVar[ConfigDict] = ConfigDict(
        allow_inf_nan=False,
        extra="forbid",
        frozen=True,
        strict=True,
    )


class TimeWindow(StrictModel):
    start_time: UtcTimestamp
    end_time: UtcTimestamp

    @model_validator(mode="after")
    def reject_reversed_window(self) -> Self:
        validate_order(self.start_time, self.end_time)
        return self


class Source(StrictModel):
    source_key: SyntheticSourceKey
    name: NonEmptyString
    kind: Literal["phone", "watch", "app", "manual"]
    bundle_id: OmittableString = None
    device_model: OmittableString = None


class HealthType(StrictModel):
    type_code: TypeCode
    display_name: NonEmptyString
    category: Literal[
        "activity",
        "blood_respiratory",
        "body",
        "environmental",
        "fitness",
        "heart",
        "other",
        "provider_specific",
        "sleep",
        "workout",
    ]
    default_unit: NonEmptyString
    sensitivity: Literal["low", "moderate", "high"]
    aliases: tuple[NonEmptyString, ...]

    @field_validator("aliases")
    @classmethod
    def reject_duplicate_aliases(
        cls,
        value: tuple[NonEmptyString, ...],
    ) -> tuple[NonEmptyString, ...]:
        if len(value) == len(set(value)):
            return value
        raise PydanticCustomError(
            ALIASES_UNIQUE_ERROR_TYPE,
            ALIASES_UNIQUE_ERROR_MESSAGE,
        )


class Sample(StrictModel):
    client_record_id: SyntheticRecordId
    source_key: SyntheticSourceKey
    type_code: TypeCode
    start_time: UtcTimestamp
    end_time: UtcTimestamp
    value: float
    unit: NonEmptyString
    metadata: dict[str, str] = Field(default_factory=dict)

    @model_validator(mode="after")
    def reject_reversed_interval(self) -> Self:
        validate_order(self.start_time, self.end_time)
        return self


class Workout(StrictModel):
    client_record_id: SyntheticRecordId
    source_key: SyntheticSourceKey
    workout_type: NonEmptyString
    start_time: UtcTimestamp
    end_time: UtcTimestamp
    duration_seconds: NonNegativeInt
    energy_kcal: OmittableNonNegativeFloat = None
    distance_meters: OmittableNonNegativeFloat = None

    @model_validator(mode="after")
    def reject_inconsistent_duration(self) -> Self:
        validate_order(self.start_time, self.end_time)
        elapsed_seconds = int(
            (
                utc_datetime(self.end_time) - utc_datetime(self.start_time)
            ).total_seconds()
        )
        if self.duration_seconds > elapsed_seconds:
            message = "duration_seconds must not exceed workout interval"
            raise ValueError(message)
        return self


class SleepStageInterval(StrictModel):
    stage: Literal["in_bed", "awake", "core", "deep", "rem"]
    start_time: UtcTimestamp
    end_time: UtcTimestamp

    @model_validator(mode="after")
    def reject_reversed_interval(self) -> Self:
        validate_order(self.start_time, self.end_time)
        return self


class SleepSession(StrictModel):
    client_record_id: SyntheticRecordId
    source_key: SyntheticSourceKey
    start_time: UtcTimestamp
    end_time: UtcTimestamp
    stage_intervals: tuple[SleepStageInterval, ...]

    @model_validator(mode="after")
    def reject_out_of_bounds_stages(self) -> Self:
        validate_order(self.start_time, self.end_time)
        session_start = utc_datetime(self.start_time)
        session_end = utc_datetime(self.end_time)
        if any(
            utc_datetime(stage.start_time) < session_start
            or utc_datetime(stage.end_time) > session_end
            for stage in self.stage_intervals
        ):
            message = "sleep stage must be contained within its session"
            raise ValueError(message)
        return self


class DeletedRecord(StrictModel):
    record_family: Literal["sample", "workout", "sleep_session"]
    source_key: SyntheticSourceKey
    client_record_id: SyntheticRecordId
    deleted_at: UtcTimestamp


class SyncCursor(StrictModel):
    source_key: SyntheticSourceKey
    cursor_kind: NonEmptyString
    cursor_value: NonEmptyString


class SyncContext(StrictModel):
    sync_window: TimeWindow
    cursors: tuple[SyncCursor, ...]


class HealthBridgeBatchV1(StrictModel):
    schema_id: Literal["health_bridge.batch.v1"]
    schema_version: str
    generated_at: UtcTimestamp
    export_window: TimeWindow
    sources: tuple[Source, ...] = Field(min_length=1)
    health_types: tuple[HealthType, ...] = Field(min_length=1)
    samples: tuple[Sample, ...]
    workouts: tuple[Workout, ...]
    sleep_sessions: tuple[SleepSession, ...]
    deleted_records: tuple[DeletedRecord, ...]
    sync: SyncContext

    @field_validator("schema_version")
    @classmethod
    def reject_unknown_major_version(cls, value: str) -> str:
        if SCHEMA_VERSION_PATTERN.fullmatch(value):
            return value
        raise PydanticCustomError(
            SCHEMA_MAJOR_ERROR_TYPE,
            SCHEMA_MAJOR_ERROR_MESSAGE,
        )

    @model_validator(mode="after")
    def reject_sync_window_outside_export_window(self) -> Self:
        export_start = utc_datetime(self.export_window.start_time)
        export_end = utc_datetime(self.export_window.end_time)
        sync_start = utc_datetime(self.sync.sync_window.start_time)
        sync_end = utc_datetime(self.sync.sync_window.end_time)
        if sync_start < export_start or sync_end > export_end:
            message = "sync window must be contained within export window"
            raise ValueError(message)
        return self

File: ios/HealthBridgeCompanion/Sources/HealthBridgeCompanionCore/BatchV1.swift
import Foundation

public struct HealthBridgeBatchV1: Codable, Equatable, Sendable {
    public let schemaID: String
    public let schemaVersion: String
    public let generatedAt: String
    public let exportWindow: HealthBridgeTimeWindow
    public let sources: [HealthBridgeSource]
    public let healthTypes: [HealthBridgeHealthType]
    public let samples: [HealthBridgeSample]
    public let workouts: [HealthBridgeWorkout]
    public let sleepSessions: [HealthBridgeSleepSession]
    public let deletedRecords: [HealthBridgeDeletedRecord]
    public let sync: HealthBridgeSyncContext

    public init(
        schemaID: String = "health_bridge.batch.v1",
        schemaVersion: String = "1.0.0",
        generatedAt: String,
        exportWindow: HealthBridgeTimeWindow,
        sources: [HealthBridgeSource],
        healthTypes: [HealthBridgeHealthType],
        samples: [HealthBridgeSample],
        workouts: [HealthBridgeWorkout],
        sleepSessions: [HealthBridgeSleepSession],
        deletedRecords: [HealthBridgeDeletedRecord],
        sync: HealthBridgeSyncContext
    ) {
        self.schemaID = schemaID
        self.schemaVersion = schemaVersion
        self.generatedAt = generatedAt
        self.exportWindow = exportWindow
        self.sources = sources
        self.healthTypes = healthTypes
        self.samples = samples
        self.workouts = workouts
        self.sleepSessions = sleepSessions
        self.deletedRecords = deletedRecords
        self.sync = sync
    }

    enum CodingKeys: String, CodingKey {
        case schemaID = "schema_id"
        case schemaVersion = "schema_version"
        case generatedAt = "generated_at"
        case exportWindow = "export_window"
        case sources
        case healthTypes = "health_types"
        case samples
        case workouts
        case sleepSessions = "sleep_sessions"
        case deletedRecords = "deleted_records"
        case sync
    }
}

public struct HealthBridgeTimeWindow: Codable, Equatable, Sendable {
    public let startTime: String
    public let endTime: String

    public init(startTime: String, endTime: String) {
        self.startTime = startTime
        self.endTime = endTime
    }

    enum CodingKeys: String, CodingKey {
        case startTime = "start_time"
        case endTime = "end_time"
    }
}

public struct HealthBridgeSource: Codable, Equatable, Sendable {
    public enum Kind: String, Codable, Equatable, Sendable {
        case phone
        case watch
        case app
        case manual
    }

    public let sourceKey: String
    public let name: String
    public let kind: Kind
    public let bundleID: String?
    public let deviceModel: String?

    public init(sourceKey: String, name: String, kind: Kind, bundleID: String? = nil, deviceModel: String? = nil) {
        self.sourceKey = sourceKey
        self.name = name
        self.kind = kind
        self.bundleID = bundleID
        self.deviceModel = deviceModel
    }

    enum CodingKeys: String, CodingKey {
        case sourceKey = "source_key"
        case name
        case kind
        case bundleID = "bundle_id"
        case deviceModel = "device_model"
    }
}

public struct HealthBridgeSample: Codable, Equatable, Sendable {
    public let clientRecordID: String
    public let sourceKey: String
    public let typeCode: String
    public let startTime: String
    public let endTime: String
    public let value: Double
    public let unit: String
    public let metadata: [String: String]

    public init(
        clientRecordID: String,
        sourceKey: String,
        typeCode: String,
        startTime: String,
        endTime: String,
        value: Double,
        unit: String,
        metadata: [String: String] = [:]
    ) {
        self.clientRecordID = clientRecordID
        self.sourceKey = sourceKey
        self.typeCode = typeCode
        self.startTime = startTime
        self.endTime = endTime
        self.value = value
        self.unit = unit
        self.metadata = metadata
    }

    enum CodingKeys: String, CodingKey {
        case clientRecordID = "client_record_id"
        case sourceKey = "source_key"
        case typeCode = "type_code"
        case startTime = "start_time"
        case endTime = "end_time"
        case value
        case unit
        case metadata
    }
}

public struct HealthBridgeWorkout: Codable, Equatable, Sendable {
    public let clientRecordID: String
    public let sourceKey: String
    public let workoutType: String
    public let startTime: String
    public let endTime: String
    public let durationSeconds: Int
    public let energyKcal: Double?
    public let distanceMeters: Double?

    public init(
        clientRecordID: String,
        sourceKey: String,
        workoutType: String,
        startTime: String,
        endTime: String,
        durationSeconds: Int,
        energyKcal: Double? = nil,
        distanceMeters: Double? = nil
    ) {
        self.clientRecordID = clientRecordID
        self.sourceKey = sourceKey
        self.workoutType = workoutType
        self.startTime = startTime
        self.endTime = endTime
        self.durationSeconds = durationSeconds
        self.energyKcal = energyKcal
        self.distanceMeters = distanceMeters
    }

    enum CodingKeys: String, CodingKey {
        case clientRecordID = "client_record_id"
        case sourceKey = "source_key"
        case workoutType = "workout_type"
        case startTime = "start_time"
        case endTime = "end_time"
        case durationSeconds = "duration_seconds"
        case energyKcal = "energy_kcal"
        case distanceMeters = "distance_meters"
    }
}

public struct HealthBridgeSleepStageInterval: Codable, Equatable, Sendable {
    public let stage: String
    public let startTime: String
    public let endTime: String

    public init(stage: String, startTime: String, endTime: String) {
        self.stage = stage
        self.startTime = startTime
        self.endTime = endTime
    }

    enum CodingKeys: String, CodingKey {
        case stage
        case startTime = "start_time"
        case endTime = "end_time"
    }
}

public struct HealthBridgeSleepSession: Codable, Equatable, Sendable {
    public let clientRecordID: String
    public let sourceKey: String
    public let startTime: String
    public let endTime: String
    public let stageIntervals: [HealthBridgeSleepStageInterval]

    public init(
        clientRecordID: String,
        sourceKey: String,
        startTime: String,
        endTime: String,
        stageIntervals: [HealthBridgeSleepStageInterval]
    ) {
        self.clientRecordID = clientRecordID
        self.sourceKey = sourceKey
        self.startTime = startTime
        self.endTime = endTime
        self.stageIntervals = stageIntervals
    }

    enum CodingKeys: String, CodingKey {
        case clientRecordID = "client_record_id"
        case sourceKey = "source_key"
        case startTime = "start_time"
        case endTime = "end_time"
        case stageIntervals = "stage_intervals"
    }
}

public struct HealthBridgeDeletedRecord: Codable, Equatable, Sendable {
    public let recordFamily: String
    public let sourceKey: String
    public let clientRecordID: String
    public let deletedAt: String

    public init(recordFamily: String, sourceKey: String, clientRecordID: String, deletedAt: String) {
        self.recordFamily = recordFamily
        self.sourceKey = sourceKey
        self.clientRecordID = clientRecordID
        self.deletedAt = deletedAt
    }

    enum CodingKeys: String, CodingKey {
        case recordFamily = "record_family"
        case sourceKey = "source_key"
        case clientRecordID = "client_record_id"
        case deletedAt = "deleted_at"
    }
}

public struct HealthBridgeSyncContext: Codable, Equatable, Sendable {
    public let syncWindow: HealthBridgeTimeWindow
    public let cursors: [HealthBridgeSyncCursor]

    public init(syncWindow: HealthBridgeTimeWindow, cursors: [HealthBridgeSyncCursor]) {
        self.syncWindow = syncWindow
        self.cursors = cursors
    }

    enum CodingKeys: String, CodingKey {
        case syncWindow = "sync_window"
        case cursors
    }
}

public struct HealthBridgeSyncCursor: Codable, Equatable, Sendable {
    public let sourceKey: String
    public let cursorKind: String
    public let cursorValue: String

    public init(sourceKey: String, cursorKind: String, cursorValue: String) {
        self.sourceKey = sourceKey
        self.cursorKind = cursorKind
        self.cursorValue = cursorValue
    }

    enum CodingKeys: String, CodingKey {
        case sourceKey = "source_key"
        case cursorKind = "cursor_kind"
        case cursorValue = "cursor_value"
    }
}
